### PR TITLE
tests: memory protect syscalls new separate stack test

### PR DESCRIPTION
**Pull request by maksimmasalski**
Add a new test test_syscall_separate_stack()

I made that test because necessary to test kernel handles system calls on a separate stack,
that can not be accessible by the calling thread. 

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

### CODE REVIEW

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 1]
```diff
@@ -112,6 +112,34 @@ static inline int z_vrfy_syscall_arg64(u64_t arg)
...
+	 */
+	int syscall_var = 10;
+	int *var_ptr = &syscall_var;
+	intptr_t ptr_int = (intptr_t)var_ptr;
+
+	struct k_thread *thread = _current;
+	unsigned int start_addr = thread->stack_info.start;
+	size_t size_addr = thread->stack_info.size;
+
+	if ((ptr_int >= (start_addr + size_addr)) || (ptr_int <= start_addr)) {
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:129]: I think this might be too loose, maybe check explicitly that the variable is inside the privileged stack?

**Review by ioannisg — COMMENTED** [review]
This seems to work for all ARCHes currently supporting user-mode - but is this a generic requirement that stack_info.start and stack_info.size define the boundaries of the user-stack (i.e. is the priv stack always beyond these boundaries)?

Other than that, just a note that a while ago I wrote a cortex-m specific test that checks exactly this thing,  thinking that this should be checked in an ARCH-specific basis

File: tests/kernel/mem_protect/syscalls/src/main.c (hunk 1 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/syscalls/src/main.c:129]: Could you send me your test name and file where it is placed in codebase? I didn't know about it.

Comment (maksimmasalski) [reply tests/kernel/mem_protect/syscalls/src/main.c:129]: How to access privileged stack?

**Comment by maksimmasalski** [comment]
I rebased that PR. @ioannisg Could you please follow up our previous conversation and answer my question?
@nashif @andyross @andrewboie Please try your best to review that PR and give comments.

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 2]
```diff
@@ -118,6 +118,34 @@ static inline int z_vrfy_syscall_arg64(uint64_t arg)
...
```
Comment (sjg20) [inline tests/kernel/mem_protect/syscalls/src/main.c:135]: Wow so many brackets!

**Review by sjg20 — APPROVED** [review]

**Comment by andrewboie** [comment]
> This seems to work for all ARCHes currently supporting user-mode - but is this a generic requirement that stack_info.start and stack_info.size define the boundaries of the user-stack (i.e. is the priv stack always beyond these boundaries)?

Yes, correct, the priv stack is never in this area. On some arches like x86 it is immediately before it.
#24714 makes this much more explicit.

File: tests/kernel/mem_protect/syscalls/src/main.c (hunk 1 above)
Comment (andrewboie) [reply tests/kernel/mem_protect/syscalls/src/main.c:135]: he wants you to validate that the stack pointer is within the bounds of the privilege stack.

we currently do not have a generic function for obtaining the privilege stack bounds. in some cases you call z_priv_stack_find() and others you do math on the base address of thread->stack_obj. it varies per architecture. can you file a bug on this and assign to me?

**Review by tyrantlizard77 — APPROVED** [review]

File: tests/kernel/mem_protect/syscalls/prj.conf [hunk 3]
```diff
@@ -3,3 +3,5 @@ CONFIG_TEST_USERSPACE=y
 CONFIG_TIMESLICING=y
 CONFIG_TIMESLICE_SIZE=10
 CONFIG_APPLICATION_DEFINED_SYSCALL=y
+CONFIG_INIT_STACKS=y
+CONFIG_THREAD_STACK_INFO=y
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/prj.conf:7]: Also, this is probably not needed, since this test requires USERSPACE, which in all ARCHes (ARC, ARM, X86) enables THREAD_STACK_INFO.

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 4]
```diff
@@ -377,6 +405,28 @@ void test_syscall_context(void)
...
+ * @brief Test to demonstrate system call is handled on a separate stack
+ *
+ * @details - Created a syscall with name test_syscall_separate_stack()
+ * - Inside of that syscall was created a local variable,
+ * and then using thread->stack_info was taken start address
+ * and size of stack of the caller thread
+ * - Then inside of the system call address of the local variable
+ * checked that it is not within the bounds of the caller thread stack
+ * - If that test is running in user mode, verify that
+ * system call is handled on a private stack,
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:418]: not sure what "private" means here, I guess the explanation in the last line is sufficient. Maybe replace private with "separate"

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 5]
```diff
@@ -377,6 +405,28 @@ void test_syscall_context(void)
...
+ * which is not accessible to the calling user mode thread.
+ *
+ * @ingroup kernel_memprotect_tests
+ */
+void test_syscall_separate_stack(void)
+{
+	zassert_true(syscall_separate_stack() == 0,
+		"System call wasn't handled by a kernel "
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:426]: by kernel

File: tests/kernel/mem_protect/syscalls/src/main.c (hunk 1 above)
Comment (ioannisg) [reply tests/kernel/mem_protect/syscalls/src/main.c:135]: @andrewboie we have arch.priv_stack_start for ARC, and ARM (not x86 AFAICS). So this could be used in this test, to add more value to the check.

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 6]
```diff
@@ -118,6 +118,34 @@ static inline int z_vrfy_syscall_arg64(uint64_t arg)
 }
 #include <syscalls/syscall_arg64_mrsh.c>
 
+ /* System call made for test test_system_call_separate_stack */
+int z_impl_syscall_separate_stack(void)
+{
+	/* Test that syscall is running on a separate stack, not accessible
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:124]: This might be a bit misleading, we do not test here that the separate stack is not accessible to the calling thread, you only test that the variable is outside the thread's normal stack.

Actually, I guess you could use the API to check that the variable is not user-accessible :)

**Review by ioannisg — CHANGES_REQUESTED** [review]
Need to clarify a few things in the comments, IMHO, and possibly add some arch-specific checks to increase the value of the test.

File: tests/kernel/mem_protect/syscalls/src/main.c (hunk 1 above)
Comment (maksimmasalski) [reply tests/kernel/mem_protect/syscalls/src/main.c:134]: @andrewboie @ioannisg I made changes for ARM and ARC, please review my new commit. As I understand there is no option for x86, and I will file an issue and assign it to Andrew to develop arch.priv_stack_start for x86.

**Comment by maksimmasalski** [comment]
@andrewboie I submitted a bug issue generic function for obtaining the privilege stack bounds https://github.com/zephyrproject-rtos/zephyr/issues/26865

File: tests/kernel/mem_protect/syscalls/src/main.c [hunk 7]
```diff
@@ -119,6 +119,45 @@ static inline int z_vrfy_syscall_arg64(uint64_t arg)
...
+	if ((ptr_int >= (start_addr + size_addr)) || (ptr_int <= start_addr)) {
+		return 0;
+	} else {
+		return -1;
+	}
+#elif defined(CONFIG_ARC) || defined(CONFIG_ARM)
+	unsigned int start_addr = thread->arch.priv_stack_start;
+
+	if ((ptr_int >= (start_addr + size_addr)) || (ptr_int <= start_addr)) {
+		return 0;
```
Comment (ioannisg) [inline tests/kernel/mem_protect/syscalls/src/main.c:144]: I suppose, return 0 is the "success" case. Then you need to return -1 here, if the variable is outside the (start, start+size)

**Comment by ioannisg** [comment]
I think this test is too arch-specific, and as such, it should be present in the ARCH test directory, for the architectures that support system calls. IIRC, the Cortex-M architecture already has a test that verifies that the system calls are executed in the privileged stack, so I guess this should only be implemented for X86 and ARC.

**Comment by maksimmasalski** [comment]
> I think this test is too arch-specific, and as such, it should be present in the ARCH test directory, for the architectures that support system calls. IIRC, the Cortex-M architecture already has a test that verifies that the system calls are executed in the privileged stack, so I guess this should only be implemented for X86 and ARC.

Understood.

**Comment by andrewboie** [comment]
> I think this test is too arch-specific, and as such, it should be present in the ARCH test directory, for the architectures that support system calls. IIRC, the Cortex-M architecture already has a test that verifies that the system calls are executed in the privileged stack, so I guess this should only be implemented for X86 and ARC.

I think if we had an arch-neutral API for getting a pointer to the privilege mode stack, then there would not need to be any arch-specific code in this test case and it could be entirely cross-platform.

Please assign a bug to me to implement this, I would rather this not be moved to arch tests.

Thanks

**Comment by andrewboie** [comment]
> I think this test is too arch-specific, and as such, it should be present in the ARCH test directory, for the architectures that support system calls. IIRC, the Cortex-M architecture already has a test that verifies that the system calls are executed in the privileged stack, so I guess this should only be implemented for X86 and ARC.

@ioannisg I think as much as possible we should not just split stuff and move it to arches.

I already have a difficult enough time getting new user mode support contributors to get all the existing tests passing, I **very** much prefer an #else/#error case if a test scenario needs an arch implementation, rather than splitting it all out and the tests never getting implemented on new user mode supporting arches at all

In other words: I want the build to break.

**Comment by maksimmasalski** [comment]
PR closed till the issue https://github.com/zephyrproject-rtos/zephyr/issues/26865 won't be closed
After solving that issue, I will reopen PR.

**Comment by maksimmasalski** [comment]
> > I think this test is too arch-specific, and as such, it should be present in the ARCH test directory, for the architectures that support system calls. IIRC, the Cortex-M architecture already has a test that verifies that the system calls are executed in the privileged stack, so I guess this should only be implemented for X86 and ARC.
> 
> I think if we had an arch-neutral API for getting a pointer to the privilege mode stack, then there would not need to be any arch-specific code in this test case and it could be entirely cross-platform.
> 
> Please assign an enhancement to me to implement this, I would rather this not be moved to arch tests.
> 
> Thanks

https://github.com/zephyrproject-rtos/zephyr/issues/26865 I already made that, is it enough? I also closed that PR till it won't be fixed.
